### ISSUE TITLE
ruby-kickstart - session2 challenge 6

### ISSUE THREAD

**Issue by ffasolin** [makersacademy/problem-solving#17]
My code is not working for this one, it seems to return true for some multiples of 3 and 7 (27, 35, 49 etc) for some reason:

```
def prime_chars? (array)
  i = array.join.length
    if i < 2
      return false
    elsif i == 2
      return true
    elsif 
      for n in (2..(i - 1))
        if i % n == 0
          return false
        else
          return true
        end
      end
    end
end
```

I already checked out the solution for this one but I don't understand why my code isn't working. Help!

**Comment by FiddlersCode** [comment]
Can you format the code?  I think if you put it in backticks it becomes formatted.  And also say what you want the code to do?

**Comment by ffasolin** [comment]
Sorry, there you go :)

**Comment by TimRobinson1** [comment]
You've got an elsif statement there that doesn't have a condition, that might be causing your issues.

**Comment by ffasolin** [comment]
Oh, true. Thanks. But changed it for else and it's still not working.

**Comment by nikeshashar** [comment]
Can you post up the error message you're getting from RSpec?

**Comment by ffasolin** [comment]
![screen shot 2017-03-23 at 11 33 21](https://cloud.githubusercontent.com/assets/25456821/24245915/c240dc34-0fbc-11e7-8a15-215e1fe84846.png)

**Comment by nikeshashar** [comment]
That should help you narrow down which line of your code is causing the problem

**Comment by ffasolin** [comment]
You mean implicitly? Because I can only see it referring to the spec file there.

**Comment by anthony-crisp** [comment]
I went about it a different way, Ruby has a Prime class that you can require before the method, that way all you have to do is count the chars and and ask if the result is .prime?.
```
require 'prime'

def prime_chars?(str)
  str.join.split("").count.prime?
end
```

It's a bit of a cheat but we have the tools built in so I figure why not use them, I don't know if this helps.

**Comment by ffasolin** [comment]
Thanks! Yes, ultimately that looks way simpler, haha. But would still like to know why mine isn't working.

**Comment by anthony-crisp** [comment]
I think it's because you are not checking if it has a whole number square root.

**Comment by Yorkshireman** [comment]
Remember when you're writing code in github you can use the little tools here to print your code nicely eg:

```
def awesome
  puts "Everyone here is amazing"
end
```

The preview tab is also helpful so you can see what it looks like before you post.

**Comment by anthony-crisp** [comment]
Thanks, will do from now on, my bad!

**Comment by AlexChalk** [comment]
```
for n in (2..(i - 1))
        if i % n == 0
          return false
        else
          return true
        end
      end
```

Hey @ffasolin, if I've understood this correctly, your code block will just run once and check whether `i % 2 == 0` . `return` breaks the code block immediately—it won't keep looping.

**Comment by AlexChalk** [comment]
tldr: I think this will work if you move `return true` to outside the loop.

The first part is correct. As soon as you get a hit for `i % n == 0`, the code block should break and report false (because one factor is enough to make a number _not_prime_). 

But just because `i % n != 0` for any one number `n`, that doesn't mean you can stop checking, as it might == 0 for some other `n`. So you have to run the loop for every number up to `i - 1`, and only then, if every number has failed the `i % n == 0` test, can you report `true`.

**Comment by ffasolin** [comment]
Yes! I think that was it. This works:
```

def prime_chars? (arr)
	i = arr.join.length
	if i < 2
		return false
	elsif i == 2
		return true
	else
		for n in 2..i-1
   			return false if i % n == 0
   			true
   		end
   	end
end
```

**Comment by FiddlersCode** [comment]
I'm working on that kata too - don't forget the absolute value issue :)

**Comment by FiddlersCode** [comment]
for an alternative solution:

```
# Test if number is prime
def isPrime(num)
  y = 2
  if num.abs < 2
    return false
  elsif num.abs == 2
    return true

  elsif num.abs > 2
    if num.abs % y == 0
      return false
    elsif
      while y < num.abs
        if num.abs % y == 0
          return false
        end
        y += 1
      end
    else
      return true
    end
  end
end
```

**Comment by anthony-crisp** [comment]
@abc17 Good spot!
My point was that he only needs to check numbers from 2 up to num's square root.

```
def prime_chars?(arr)
  num = arr.join.chars.count
  return false if num <= 1
  (2..Math.sqrt(num)).each { |i| return false if num % i == 0  }
  true
end
```

But you spotted the bug in the code as it was and so answered op's question.

**Comment by AlexChalk** [comment]
Ha @anthony-crisp , you're better at maths then me. On reflection, I guess you could stop at num/2, but I don't know enough maths to explain your version.

**Comment by anthony-crisp** [comment]
This explains it much better than I can:

[http://stackoverflow.com/questions/5811151/why-do-we-check-up-to-the-square-root-of-a-prime-number-to-determine-if-it-is-prime](url)

**Comment by anthony-crisp** [comment]
But basically if a number has a whole number square root it cant be prime, if its divisible by any of the numbers lower than its square root its not prime, any number after its square root would be irrelevant as it would have to be multiplied by one of the numbers below the square root to equal the number, so you only have to check if it mod any number below the square root equals zero.

**Comment by AlexChalk** [comment]
Thanks, I think that makes sense. Your link is broken though: http://stackoverflow.com/questions/5811151/why-do-we-check-up-to-the-square-root-of-a-prime-number-to-determine-if-it-is-prime

**Comment by anthony-crisp** [comment]
Sorry, thanks for the re-post

**Comment by anthony-crisp** [comment]
Anyway, its easier to just require 'prime' and let Ruby do the thinking for you ;-)
I think we can close this now?
